File: notification-service/src/main/java/com/mealmap/notificationservice/core/dto/filter/NotificationFilter.java
package com.mealmap.notificationservice.core.dto.filter;

import com.mealmap.notificationservice.document.enums.Channel;
import com.mealmap.notificationservice.document.enums.NotificationStatus;
import jakarta.validation.constraints.AssertTrue;
import jakarta.validation.constraints.PastOrPresent;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;

import java.time.LocalDateTime;

import static com.mealmap.notificationservice.validator.RangeValidator.isValidRange;

@Data
@Builder
@AllArgsConstructor
public class NotificationFilter {
    private Channel channel;

    private NotificationStatus status;

    @PastOrPresent
    private LocalDateTime sentAtStart;

    @PastOrPresent
    private LocalDateTime sentAtEnd;

    @AssertTrue(message = "sentAtStart must be before sentAtEnd")
    private boolean isValidOrderedAtRange() {
        return isValidRange(sentAtStart, sentAtEnd);
    }
}

File: frontend/src/api/notification/dto/NotificationFilter.ts
import { Channel } from "../enums/Channel";
import { NotificationStatus } from "../enums/NotificationStatus";

export interface NotificationFilter {
  userId?: string; // UUID as a string
  channel?: Channel;
  status?: NotificationStatus;
  sentAtStart?: string; // ISO date-time string
  sentAtEnd?: string; // ISO date-time string
}
